bundles/org.openhab.binding.warmup/src/main/java/org/openhab/binding/warmup/internal/handler/RoomHandler.java

```diff
@@ -0,0 +1,147 @@
+import org.openhab.core.library.types.QuantityType;
+import org.openhab.core.library.unit.SIUnits;
+import org.openhab.core.thing.ChannelUID;
+import org.openhab.core.thing.Thing;
+import org.openhab.core.thing.ThingStatus;
+import org.openhab.core.thing.ThingStatusDetail;
+import org.openhab.core.types.Command;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+/**
+ * @author James Melville - Initial contribution
+ */
+@NonNullByDefault
+public class RoomHandler extends WarmupThingHandler implements WarmupRefreshListener {
+
+    private final Logger logger = LoggerFactory.getLogger(RoomHandler.class);
+    private @Nullable RoomConfigurationDTO config;
+
+    public RoomHandler(Thing thing) {
+        super(thing);
+    }
+
+    @Override
+    public void initialize() {
+        super.initialize();
+        config = getConfigAs(RoomConfigurationDTO.class);
+        super.refreshFromServer();
+    }
+
+    @Override
+    public void handleCommand(ChannelUID channelUID, Command command) {
+        super.handleCommand(channelUID, command);
+        if (CHANNEL_TARGET_TEMPERATURE.equals(channelUID.getId()) && command instanceof QuantityType<?>) {
+            setOverride((QuantityType<?>) command);
+        }
+        if (CHANNEL_FROST_PROTECTION_MODE.equals(channelUID.getId()) && command instanceof OnOffType) {
+            toggleFrostProtectionMode((OnOffType) command);
+        }
+    }
+
+    /**
+     * Process device list and populate room properties, status and state
+     *
+     * @param domain Data model representing all devices
+     */
+    @Override
+    public void refresh(@Nullable QueryResponseDTO domain) {
+        if (domain == null) {
+            updateStatus(ThingStatus.OFFLINE, ThingStatusDetail.COMMUNICATION_ERROR, "No data from bridge");
+        } else if (config != null) {
+            final String serialNumber = config.getSerialNumber();
+            for (LocationDTO location : domain.getData().getUser().getLocations()) {
+                for (RoomDTO room : location.getRooms()) {
+                    if (room.getThermostat4ies() != null && !room.getThermostat4ies().isEmpty()
+                            && room.getThermostat4ies().get(0).getDeviceSN().equals(serialNumber)) {
+                        if (room.getThermostat4ies().get(0).getLastPoll() > 10) {
+                            updateStatus(ThingStatus.OFFLINE, ThingStatusDetail.COMMUNICATION_ERROR,
+                                    "Thermostat has not polled for 10 minutes");
+                        } else {
+                            updateStatus(ThingStatus.ONLINE);
+
+                            updateProperty("Id", room.getId());
+                            updateProperty("Serial Number", serialNumber);
+                            updateProperty("Name", room.getName());
+                            updateProperty("Location Id", location.getId());
+                            updateProperty("Location", location.getName());
```

Comment:
As you use these properties in the discovery and below too, you could make the names constants.